### PR TITLE
[vm file format] Use a macro to derive the number of bytecode instructions

### PR DESCRIPTION
**Pull request by sunshowers**
My first proc macro! Decided to tackle this easy one.

### CODE REVIEW

File: common/num-variants/tests/basic.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+#[test]
+fn zero_enum() {
+    assert_eq!(ZeroEnum::NUM_VARIANTS, 0);
+}
+
+#[test]
+fn custom_name() {
+    assert_eq!(CustomName::CUSTOM_NAME, 3);
+}
```
Comment (tzakian) [inline common/num-variants/tests/basic.rs]: Would it be possible to add some negative tests here (i.e. trying to derive this on a struct or union)? Not sure if that's possible though due to compile-time stuff and testing.

Comment (sunshowers) [reply common/num-variants/tests/basic.rs]: Yeah it's possible to do with compiletest-rs. I don't have the time for it atm but I'll add a TODO.

Comment (dtolnay) [reply common/num-variants/tests/basic.rs]: You want https://github.com/dtolnay/trybuild, compiletest-rs is awful.

**Pull request closed without merging**